### PR TITLE
[WIP][Feature] Support SOT method: STARK

### PR DESCRIPTION
**Pull request by JingweiZhang12**
The performance of STARK is aligned with the paper.
Therefore, the STARK algorithm will be merged into master in December 2021.

### CODE REVIEW

File: mmtrack/models/sot/stark.py [hunk 1]
```diff
@@ -0,0 +1,345 @@
...
+
+        att_mask = cv2.resize(att_mask, (output_size, output_size)).astype(np.bool_)
+
+        return img_crop_padded, resize_factor, att_mask
+
+    def _normalize(self, img_tensor, amask_arr):
+        self.mean = torch.tensor([0.485, 0.456, 0.406]).view(
+            (1, 3, 1, 1)).cuda()
+        self.std = torch.tensor([0.229, 0.224, 0.225]).view(
+            (1, 3, 1, 1)).cuda()
```
Comment (GT9505) [inline mmtrack/models/sot/stark.py:154]: refer [here](https://github.com/open-mmlab/mmtracking/blob/master/mmtrack/core/motion/flow.py#L27) to change code

File: mmtrack/models/track_heads/stark_head.py [hunk 2]
```diff
@@ -0,0 +1,376 @@
...
+                norm_cfg=dict(type='BN', requires_grad=True),
+                act_cfg=dict(type='ReLU'),
+                inplace=True)
+
+        # top-left corner
+        self.conv1_tl = conv_module(inplanes, channel)
+        self.conv2_tl = conv_module(channel, channel // 2)
+        self.conv3_tl = conv_module(channel // 2, channel // 4)
+        self.conv4_tl = conv_module(channel // 4, channel // 8)
+        self.conv5_tl = nn.Conv2d(channel // 8, 1, kernel_size=1)
```
Comment (GT9505) [inline mmtrack/models/track_heads/stark_head.py:52]: use ConvModule

File: mmtrack/models/track_heads/stark_head.py [hunk 3]
```diff
@@ -0,0 +1,376 @@
...
+            x (Tensor): of shape [bs, C, H, W].
+        Returns:
+            score_map_tl (Tensor[bs, 1, H, W]): the score map of top left corner of tracking bbox
+            score_map_br (Tensor[bs, 1, H, W]): the score map of bottom right corner of tracking bbox
+        """
+        # top-left branch
+        x_tl1 = self.conv1_tl(x)
+        x_tl2 = self.conv2_tl(x_tl1)
+        x_tl3 = self.conv3_tl(x_tl2)
+        x_tl4 = self.conv4_tl(x_tl3)
```
Comment (GT9505) [inline mmtrack/models/track_heads/stark_head.py:94]: use module_list

File: mmtrack/models/track_heads/stark_head.py [hunk 4]
```diff
@@ -0,0 +1,376 @@
...
+from mmcv.cnn.bricks import ConvModule
+from mmdet.core.bbox.transforms import bbox_xyxy_to_cxcywh
+from mmdet.models import HEADS
+from mmdet.models.dense_heads import DETRHead
+from mmdet.models.utils import Transformer
+from mmdet.models.utils.builder import TRANSFORMER
+from torch import nn
+
+@HEADS.register_module()
+class CornerPredictorHead(nn.Module):
```
Comment (GT9505) [inline mmtrack/models/track_heads/stark_head.py:12]: use BaseModule

File: mmtrack/models/track_heads/stark_head.py [hunk 5]
```diff
@@ -0,0 +1,376 @@
...
+    def __init__(self,
+                 input_dim,
+                 hidden_dim,
+                 output_dim,
+                 num_layers,
+                 BN=False):
+        super().__init__()
+        self.num_layers = num_layers
+        hidden_dims = [hidden_dim] * (num_layers - 1)
+        if BN:
```
Comment (GT9505) [inline mmtrack/models/track_heads/stark_head.py:141]: change `BN` to `norm_cfg`

File: mmtrack/models/track_heads/stark_head.py [hunk 6]
```diff
@@ -0,0 +1,376 @@
...
+        )
+
+    def _init_layers(self):
+        feat_size = self.test_cfg['search_size'] // self.stride
+        self.bbox_head = CornerPredictorHead(
+            self.embed_dims,
+            self.embed_dims,
+            feat_size=feat_size,
+            stride=self.stride)
+        self.cls_head = ScoreHead(self.embed_dims, self.embed_dims, 1,
```
Comment (GT9505) [inline mmtrack/models/track_heads/stark_head.py:276]: for the names of `CornerPredictorHead` and `ScoreHead`, do we have names mentioned in the paper?

File: configs/sot/stark/stark.py [hunk 7]
```diff
@@ -0,0 +1,98 @@
...
+                        num_heads=8,
+                        attn_drop=0.1,
+                        dropout_layer=dict(type='Dropout', drop_prob=0.1)),
+                    feedforward_channels=2048,
+                    ffn_cfgs=dict(embed_dims=256, ffn_drop=0.1),
+                    operation_order=('self_attn', 'norm', 'cross_attn', 'norm',
+                                     'ffn', 'norm'))),
+        ),
+        positional_encoding=dict(
+            type='SinePositionalEncoding', num_feats=128, normalize=True)),
```
Comment (GT9505) [inline configs/sot/stark/stark.py:62]: we need to show the score_head and regression_head in the config

File: mmtrack/models/track_heads/stark_head.py (hunk 2 above)
Comment (JingweiZhang12) [reply mmtrack/models/track_heads/stark_head.py:52]: use ConvModule is like this: 
ConvModule(
                in_channels=in_planes,
                out_channels=out_planes,
                kernel_size=kernel_size,
                padding=padding,
                bias=True,
                norm_cfg=dict(type='BN', requires_grad=True)
then use nn.Sequential or nn.ModuleList to warp five that ConvModule may seem ugly.

File: mmtrack/models/track_heads/stark_head.py (hunk 6 above)
Comment (JingweiZhang12) [reply mmtrack/models/track_heads/stark_head.py:276]: not explicitly.   mention "bbox prediction head" and "score prediction head"

File: configs/sot/stark/stark_st1_r50_1x_got10k.py [hunk 8]
```diff
@@ -0,0 +1,218 @@
...
+    type='Stark',
+    backbone=dict(
+        type='ResNet',
+        depth=50,
+        num_stages=3,
+        strides=(1, 2, 2),
+        dilations=[1, 1, 1],
+        out_indices=[2],
+        frozen_stages=1,
+        norm_eval=True,
```
Comment (JingweiZhang12) [inline configs/sot/stark/stark_st1_r50_1x_got10k.py:16]:
```suggestion
        out_indices=[2],
        frozen_stages=1,
        norm_eval=False,
```

**Comment by ly59782** [comment]
> The performance of STARK is aligned with the paper. Therefore, the STARK algorithm will be merged into master in December 2021.

hi, thans for your work , would you provide the checkpoints file? many thanks!!

**Comment by GT9505** [comment]
> > The performance of STARK is aligned with the paper. Therefore, the STARK algorithm will be merged into master in December 2021.
> 
> hi, thans for your work , would you provide the checkpoints file? many thanks!!

@JingweiZhang12

**Comment by GT9505** [comment]
@ly59782 The STARK method has not been merged into master yet. We will provide the ckpt once the method is merged.

**Comment by GT9505** [comment]
The pretrained models of STARK is released in #443 .Therefore I will close this PR.
